module/zfs/zfs_vfsops.c

```diff
@@ -972,6 +972,152 @@ zfs_id_overquota(zfsvfs_t *zfsvfs, uint64_t usedobj, uint64_t id)
 	    zfs_id_overobjquota(zfsvfs, usedobj, id));
 }
 
+static zfsvfs_kstats_t empty_zfsvfs_kstats = {
+	{ "writes",	KSTAT_DATA_UINT64 },
+	{ "nwritten",	KSTAT_DATA_UINT64 },
+	{ "reads",	KSTAT_DATA_UINT64 },
+	{ "nread",	KSTAT_DATA_UINT64 },
+};
+
+static int
+zfsvfs_kstats_update(kstat_t *ksp, int rw)
+{
+	zfsvfs_t *zfsvfs = ksp->ks_private;
+	zfsvfs_kstats_t *zs = ksp->ks_data;
+
+	ASSERT3P(zfsvfs->z_iokstat->ks_data, ==, zs);
+
+	if (rw == KSTAT_WRITE)
+		return (EACCES);
+
+	zs->zk_writes.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_writes);
+	zs->zk_nwritten.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_nwritten);
+	zs->zk_reads.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_reads);
+	zs->zk_nread.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_nread);
+
+	return (0);
+}
+
+static void
+zfsvfs_kstats_create(zfsvfs_t *zfsvfs)
+{
+	ASSERT3P(zfsvfs->z_iokstat, ==, NULL);
+
+	/*
+	 * There should not be anything wrong with having kstats for
+	 * snapshots. Since we are not sure how useful they would be
+	 * though nor how much their memory overhead would matter in
+	 * a filesystem with many snapshots, we skip them for now.
+	 */
+	if (zfsvfs->z_issnap)
+		return;
+
+	/*
+	 * We want the code and the behavior of ZFS to be consistent and
+	 * uniform between different OSes that ZFS runs on. Thus for the
+	 * name of this kstat we are limited by KSTAT_STRLEN as definied
+	 * in illumos which is less than what we have in Linux
+	 * (31 vs 255 characters).
+	 *
+	 * Thus, to keep things the same we distinguish handles by their
+	 * objset IDs prepended by the pool's load_guid. Note that both
+	 * numbers are formatted in hex to save character space.
+	 * See breakdown in the next comment.
+	 *
+	 * We use the pool's load_guid because it is guaranteed to not
+	 * change as long as the machine is running (unlike the current
+	 * GUID from the pool's config which could change upona reguid).
+	 */
+	char kstat_name[KSTAT_STRLEN];
+	int n = snprintf(kstat_name, sizeof (kstat_name), "%llx-%llx",
```

Comment:
In Linux, unlike illumos, there is already a subdirectory for the pool kstats, by name. This works in Linux because the kstat namespace is heirarchical, not flat like illumos. Rather than using the load guid here, let's use the pool's subdirectory.
I'm thinking something like:
```
int n = snprintf(kstat_name, sizeof (kstat_name), "%s/objset-0x%llx", 
     spa_name(dmu_objset_spa(zfsvfs->z_os)),
     (unsigned long long)dmu_objset_id(zfsvfs->z_os));
```

we might be able to get more clever with the objset name, too

Reply:
Yes. If you read my comment in the code I do mention that (and I clarify more on my other comment on github). The reason why I do this is to keep the codebase and the format of the key for retrieval the same between OSs.

Reply:
I'm open to changing this to something else than what I have right now. That said, I think it would be easier to do this for now from the perspective of porting. Plus in the future if some utility ships in the zol userland that is ported to illumos and FreeBSD, the routines for formatting the key for the kstat retrieval will be the same across all platforms.

Reply:
But it (kstats) already isn't the same in Linux. There is no kstat(1), so there isn't a way other than traversing the procfs tree to get there. Use the tree, Luke!

Reply:
> We want the code and the behavior of ZFS to be consistent and uniform between different OSes 

Normally yes, but in this case I'm not convinced it's all that important.  The way kstats are accessed on Linux is different enough from the other platforms that I don't really see any benefit from inheriting these old limitations.  Any code that consumes the kstats is already going to need to be Linux specific _unless_ someone ports the userspace kstat libraries to Linux too.

I think adopting @richardelling's proposed naming scheme for now would be reasonable.  Since pools can't currently be renamed without re-importing the pool name is functionally as good as the load_guid.  If we wanted to get fancier creating a subdirectory for each objset id would be nice, we're going to inevitably want to add additional kstats per-dataset.

Reply:
ok. I'll change the current design to the above naming scheme.

Reply:
I'd suggest preserving everything but the final name to make it easy for the other OpenZFS implementations to use.

Reply:
Review update changed the naming scheme to the one proposed above.